File: Diagrammatize.Design.Pattern/code/python/Interpreter/Sample2/CommandListNode.py

import io
from Node import Node
from ParseException import ParseException

# <command list> ::= <command>* end
class CommandListNode(Node):
    def __init__(self):
        super().__init__()
        self.list = []

    def parse(self, context):
        from CommandNode import CommandNode
        while True:
            if not context.currentToken():
                raise ParseException("Missing 'end'")
            elif context.currentToken() == "end":
                context.skipToken("end")
                break
            else:
                commandNode = CommandNode()
                commandNode.parse(context)
                self.list.append(commandNode)

    def __str__(self):
        str_list = [str(token) for token in self.list]
        buffer = io.StringIO()
        buffer.write("[")
        buffer.write(" ".join(str_list))
        buffer.write("]")
        return buffer.getvalue()

File: Diagrammatize.Design.Patterns/code/cxx/Interpreter/Sample2/CommandListNode.cpp
#include "CommandListNode.hpp"

#include <sstream>

#include "CommandNode.hpp"
#include "ParseException.hpp"

CommandListNode::~CommandListNode() {}

void CommandListNode::parse(Context& context) {
    while (true) {
        if (context.currentToken().empty()) {
            throw ParseException("Missing 'end'");
        } else if (context.currentToken() == "end") {
            context.skipToken("end");
            break;
        } else {
            auto commandNode = std::unique_ptr<Node>(new CommandNode());
            commandNode->parse(context);
            list_.push_back(std::move(commandNode));
        }
    }
}

std::string CommandListNode::toString() {
    int n = list_.size();
    std::stringstream ss;
    ss << "[";
    for (int i = 0; i < n; i++) {
        ss << list_[i]->toString();
        if (i != n-1) {
            ss << " ";
        }
    }
    ss << "]";

    return ss.str();
}
